tools/lsp/common/rename_component.rs

```diff
@@ -0,0 +1,781 @@
+    old_type: &str,
+    new_type: &str,
+    edits: &mut Vec<(SourceFile, lsp_types::TextEdit)>,
+) {
+    for component in document_node.Component() {
+        replace_element_types(&component.Element(), old_type, new_type, edits);
+    }
+    for exported in document_node.ExportsList() {
+        if let Some(component) = exported.Component() {
+            replace_element_types(&component.Element(), old_type, new_type, edits);
+        }
+    }
+}
+
+fn fix_exports(
+    type_loader: &TypeLoader,
+    document_node: &syntax_nodes::Document,
+    old_type: &str,
+    new_type: &str,
+    edits: &mut Vec<(SourceFile, lsp_types::TextEdit)>,
+) {
+    for export in document_node.ExportsList() {
+        for specifier in export.ExportSpecifier() {
+            let identifier = specifier.ExportIdentifier();
+            if identifier.text().to_string().trim() == old_type {
+                let Some(source_file) = identifier.source_file() else {
+                    continue;
+                };
+                edits.push((
+                    source_file.clone(),
+                    lsp_types::TextEdit {
+                        range: util::map_node(&identifier)
+                            .expect("This needs to have a source file"),
+                        new_text: new_type.to_string(),
+                    },
+                ));
+
+                let update_imports = if let Some(export_name) = specifier.ExportName() {
+                    // Remove "as Foo"
+                    if export_name.text().to_string().trim() == new_type {
+                        let start_position =
+                            util::map_position(&source_file, identifier.text_range().end());
+                        let end_position =
+                            util::map_position(&source_file, export_name.text_range().end());
+                        edits.push((
+                            source_file.clone(),
+                            lsp_types::TextEdit {
+                                range: lsp_types::Range::new(start_position, end_position),
+                                new_text: String::new(),
+                            },
+                        ));
+                        true
+                    } else {
+                        false
+                    }
+                } else {
+                    true
+                };
+
+                if update_imports {
+                    let my_path = document_node.source_file.path().to_owned();
+                    fix_imports(type_loader, &my_path, old_type, new_type, edits);
+                }
+            }
+        }
+    }
+}
+
+pub fn rename_component_from_definition(
+    type_loader: &TypeLoader,
+    identifier: &syntax_nodes::DeclaredIdentifier,
+    new_name: String,
+) -> crate::Result<lsp_types::WorkspaceEdit> {
+    let Some(source_file) = identifier.source_file() else {
+        return Err("identifier has no associated source file".into());
```

Comment:
There need to be a bit of polish for the error messages in this function.
Some uniformity (they should all start with an uppercase) and be a bit more descriptive: "This component cannot be renamed because ..."

If this is really some error that cannot happen, then it should be a panic. Or some "Internal Error" kind of error.

But remember this is user-facing error message so it needs to be understandable and actionable for the user.